File: backend/infrastructure/scheduler.py
"""
Background scheduler for autonomous agent chat rounds.

This module runs periodic tasks to process agent conversations in active rooms,
enabling background chatroom interactions when users are not actively viewing.

Architecture note:
- The scheduler identifies active rooms that need processing
- It delegates actual orchestration to ChatOrchestrator.process_autonomous_round()
- Game rooms (rooms with world_id) are excluded - they use TRPGOrchestrator instead
"""

import asyncio
import logging
from contextlib import asynccontextmanager
from datetime import datetime, timedelta, timezone

from apscheduler.schedulers.asyncio import AsyncIOScheduler
from orchestration import ChatOrchestrator
from sdk import AgentManager
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from sqlalchemy.orm import selectinload

from infrastructure.database import models

logger = logging.getLogger("BackgroundScheduler")

# Suppress noisy APScheduler "max instances reached" warnings
# These are expected during heavy load and not actionable
logging.getLogger("apscheduler.scheduler").setLevel(logging.ERROR)


class BackgroundScheduler:
    """Manages background tasks for autonomous agent conversations."""

    def __init__(
        self,
        chat_orchestrator: ChatOrchestrator,
        agent_manager: AgentManager,
        get_db_session,
        max_concurrent_rooms: int = 5,
    ):
        self.scheduler = AsyncIOScheduler()
        self.chat_orchestrator = chat_orchestrator
        self.agent_manager = agent_manager
        self.get_db_session = get_db_session
        self.max_concurrent_rooms = max_concurrent_rooms
        self.is_running = False
        # Create semaphore once to properly enforce concurrent room limit
        self._room_semaphore = asyncio.Semaphore(max_concurrent_rooms) if max_concurrent_rooms else None

    def start(self):
        """Start the background scheduler."""
        if not self.is_running:
            # Run autonomous chat rounds every 2 seconds
            self.scheduler.add_job(
                self._process_active_rooms,
                "interval",
                seconds=2,
                id="process_active_rooms",
                replace_existing=True,
                max_instances=1,  # Only one instance at a time (prevents overwhelming system)
                coalesce=True,  # Skip missed runs if previous job still running
                misfire_grace_time=None,  # Never misfire - just skip if busy (suppresses warnings)
            )

            # Clean up expired cache entries every 5 minutes
            self.scheduler.add_job(
                self._cleanup_cache, "interval", minutes=5, id="cleanup_cache", replace_existing=True
            )

            self.scheduler.start()
            self.is_running = True
            logger.info(
                "🚀 Background scheduler started - processing rooms every 2 seconds, cache cleanup every 5 minutes"
            )

    def stop(self):
        """Stop the background scheduler."""
        if self.is_running:
            self.scheduler.shutdown()
            self.is_running = False
            logger.info("🛑 Background scheduler stopped")

    async def _process_active_rooms(self):
        """
        Process autonomous chat rounds for all active rooms.

        A room is considered active if:
        - It has messages in the last 5 minutes
        - It's not paused
        - It has at least 2 agents
        """
        try:
            async with self._session_scope() as db:
                active_rooms = await self._get_active_rooms(db)

            if not active_rooms:
                # Don't log when there's no activity (too noisy)
                return

            logger.info(f"🔄 Processing {len(active_rooms)} active room(s)")

            async def process_with_error_handling(room):
                try:
                    if self._room_semaphore:
                        async with self._room_semaphore:
                            await self._process_room_for_background_job(room)
                    else:
                        await self._process_room_for_background_job(room)
                except Exception as e:
                    logger.error(f"❌ Error processing room {room.id}: {e}")
                    import traceback

                    traceback.print_exc()

            # Process all active rooms concurrently with a small cap
            await asyncio.gather(*[process_with_error_handling(room) for room in active_rooms])

        except Exception as e:
            logger.error(f"💥 Error in _process_active_rooms: {e}")
            import traceback

            traceback.print_exc()

    @asynccontextmanager
    async def _session_scope(self):
        """
        Provides a database session with proper transaction handling.

        - Commits on successful completion
        - Rolls back on any exception
        - Always closes the session properly
        """
        session_gen = self.get_db_session()
        session = await anext(session_gen)
        try:
            yield session
            await session.commit()  # Explicit commit on success
        except Exception:
            await session.rollback()  # Explicit rollback on error
            raise
        finally:
            try:
                await session_gen.aclose()
            except Exception as e:
                logger.error(f"Error closing database session: {e}")

    async def _process_room_for_background_job(self, room: models.Room):
        async with self._session_scope() as room_db:
            await self._process_room_autonomous_round(room_db, room)

    async def _get_active_rooms(self, db: AsyncSession) -> list:
        """
        Get rooms that should have autonomous agent interactions.

        Criteria:
        - Has messages in the last 5 minutes
        - Not paused
        - Not finished (all agents haven't skipped)
        - Has at least 2 agents
        """
        # Calculate cutoff time (5 minutes ago)
        cutoff_time = datetime.now(timezone.utc) - timedelta(minutes=5)

        # Use the room's last_activity_at field to avoid repeated full message scans
        # Exclude game rooms (rooms with world_id) - they use TRPGOrchestrator, not ChatOrchestrator
        stmt = (
            select(models.Room)
            .options(selectinload(models.Room.agents))  # Eager load agents
            .where(
                models.Room.is_paused == False,
                models.Room.is_finished == False,
                models.Room.last_activity_at >= cutoff_time,
                models.Room.world_id.is_(None),  # Exclude TRPG game rooms
            )
            .order_by(models.Room.last_activity_at.desc())
        )

        # Optionally cap the number of rooms fetched to reduce load during spikes
        if self.max_concurrent_rooms:
            stmt = stmt.limit(self.max_concurrent_rooms)

        result = await db.execute(stmt)
        rooms = result.scalars().all()

        # Filter rooms with at least 2 agents (agents already loaded via selectinload)
        active_rooms = [room for room in rooms if len(room.agents) >= 2]

        return active_rooms

    def _cleanup_completed_tasks(self):
        """Remove completed tasks from active_room_tasks to prevent memory leak."""
        completed_rooms = [room_id for room_id, task in self.chat_orchestrator.active_room_tasks.items() if task.done()]
        for room_id in completed_rooms:
            del self.chat_orchestrator.active_room_tasks[room_id]
            logger.debug(f"Cleaned up completed task for room {room_id}")

    async def _process_room_autonomous_round(self, db, room: models.Room):
        """
        Process one autonomous round for a room.

        Delegates to ChatOrchestrator.process_autonomous_round() which handles
        the tape-based scheduling logic. This keeps orchestration logic centralized.
        """
        # Clean up completed tasks before processing
        self._cleanup_completed_tasks()

        # Delegate to ChatOrchestrator for the actual orchestration
        await self.chat_orchestrator.process_autonomous_round(
            db=db,
            room=room,
            agent_manager=self.agent_manager,
        )

    async def _cleanup_cache(self):
        """
        Clean up expired cache entries and stale orchestrator/agent state.
        This runs every 5 minutes to prevent memory bloat.
        """
        try:
            from infrastructure.cache import get_cache

            cache = get_cache()
            cache.cleanup_expired()
            cache.log_stats()

            # Clean up stale orchestrator entries to prevent memory leaks
            self.chat_orchestrator.cleanup_stale_entries()

            # Clean up stale agent manager resources (task locks, etc.)
            self.agent_manager.cleanup_stale_resources()
        except Exception as e:
            logger.error(f"Error during cache cleanup: {e}")

File: backend/src/infrastructure/scheduler.ts
/**
 * Background scheduler for autonomous agent chat rounds: every **2 seconds**
 * each active chat room gets one round, so the agents keep talking while nobody
 * is watching; every **5 minutes** expired cache entries and stale orchestrator
 * state are swept. It picks rooms and does not orchestrate — *how* a round runs
 * belongs to {@link RoomOrchestrator.handleAutonomousRound}, which gives a
 * background round the interrupt and in-flight rules a user message gets.
 *
 * **A tick arriving while the previous one is still running is dropped, not
 * queued.** A bare `setInterval` with an async callback does the opposite: a
 * slow tick stacks up overlapping runs that all hit the same rooms.
 * {@link BackgroundScheduler.tick} guards on the in-flight promise instead.
 *
 * The active-room select lives here rather than in `src/crud/`: it is the
 * scheduler's own selection policy and exists nowhere else.
 */

import { and, count, desc, eq, gte, inArray, isNull } from 'drizzle-orm'

import type { Db } from '../db'
import { roomAgents, rooms, type Room } from '../db/schema'
import { getCache } from './cache'
import { getLogger } from './logging/logger'

const logger = getLogger('BackgroundScheduler')

export const PROCESS_INTERVAL_MS = 2_000

export const CLEANUP_INTERVAL_MS = 5 * 60_000

/**
 * How recently a room must have seen activity to be scheduled. Read off
 * `rooms.last_activity_at` rather than by scanning messages: the column exists
 * so this poll does not re-count a room's whole transcript every two seconds.
 */
export const ACTIVE_WINDOW_MS = 5 * 60_000

/** Below this, there is nobody for an agent to talk *to*. */
const MIN_AGENTS = 2

/** Narrowed to two methods, so a test can drive the scheduler without a session
 * pool or a model and this module declares its own dependency shape. */
export interface SchedulerOrchestrator {
  handleAutonomousRound(roomId: number): Promise<{ skipped?: boolean }>
  cleanupStaleEntries?(maxAgeSeconds?: number): number
}

export interface BackgroundSchedulerDeps {
  db: Db
  orchestrator: SchedulerOrchestrator
  /** Caps the rooms selected per tick (`LIMIT`) *and* how many run at once (a
   * semaphore) — the latter also covers a tick slower than the interval. */
  maxConcurrentRooms: number
  /** Overridable so tests do not have to wait on real wall-clock intervals. */
  processIntervalMs?: number
  cleanupIntervalMs?: number
}

export class BackgroundScheduler {
  private readonly db: Db
  private readonly orchestrator: SchedulerOrchestrator
  private readonly maxConcurrentRooms: number
  private readonly processIntervalMs: number
  private readonly cleanupIntervalMs: number

  private processTimer: ReturnType<typeof setInterval> | null = null
  private cleanupTimer: ReturnType<typeof setInterval> | null = null

  /** The in-flight tick, or null. Non-null is what makes the next tick a no-op. */
  private inFlight: Promise<void> | null = null
  private running = false

  constructor(deps: BackgroundSchedulerDeps) {
    this.db = deps.db
    this.orchestrator = deps.orchestrator
    this.maxConcurrentRooms = deps.maxConcurrentRooms
    this.processIntervalMs = deps.processIntervalMs ?? PROCESS_INTERVAL_MS
    this.cleanupIntervalMs = deps.cleanupIntervalMs ?? CLEANUP_INTERVAL_MS
  }

  get isRunning(): boolean {
    return this.running
  }

  /** Idempotent. */
  start(): void {
    if (this.running) return
    this.running = true

    this.processTimer = setInterval(() => void this.tick(), this.processIntervalMs)
    this.cleanupTimer = setInterval(() => {
      this.cleanup()
    }, this.cleanupIntervalMs)

    // A pending timer must not be the reason the process stays alive: shutdown
    // is driven by the server, and an un-unref'd interval would outlive it.
    this.processTimer.unref?.()
    this.cleanupTimer.unref?.()

    logger.info(
      `🚀 Background scheduler started - processing rooms every ${this.processIntervalMs / 1000}s, ` +
        `cache cleanup every ${this.cleanupIntervalMs / 60_000} minutes`,
    )
  }

  /**
   * Stop the timers and wait for the in-flight tick. Idempotent, and awaitable
   * because a half-finished tick holds a database handle a test is about to
   * delete — shutdown must join it, not merely stop scheduling more.
   */
  async stop(): Promise<void> {
    if (this.processTimer) clearInterval(this.processTimer)
    if (this.cleanupTimer) clearInterval(this.cleanupTimer)
    this.processTimer = null
    this.cleanupTimer = null

    const inFlight = this.inFlight
    if (!this.running && inFlight === null) return

    this.running = false
    // `tick` never rejects, so this cannot poison shutdown.
    if (inFlight) await inFlight

    logger.info('🛑 Background scheduler stopped')
  }

  /**
   * One pass over the active rooms; public so tests can drive it directly. Never
   * rejects and never wedges: one room's failure is logged and the tick
   * continues, and a failure of the tick itself still clears {@link inFlight}.
   */
  async tick(): Promise<void> {
    // Skip, do not queue — overlapping ticks would hit the same rooms twice.
    if (this.inFlight !== null) {
      logger.debug('Previous tick still running, skipping this one')
      return
    }

    const run = this.processActiveRooms().catch((error: unknown) => {
      logger.exception('💥 Error in background tick', error)
    })
    this.inFlight = run

    try {
      await run
    } finally {
      this.inFlight = null
    }
  }

  private async processActiveRooms(): Promise<void> {
    const active = this.activeRooms()
    // Silent when there is nothing to do: this runs every two seconds.
    if (active.length === 0) return

    logger.info(`🔄 Processing ${active.length} active room(s)`)

    const limit = semaphore(this.maxConcurrentRooms)
    await Promise.all(
      active.map((room) =>
        limit(async () => {
          try {
            await this.orchestrator.handleAutonomousRound(room.id)
          } catch (error: unknown) {
            // One bad room must not take the tick down with it.
            logger.exception(`❌ Error processing room ${room.id}`, error)
          }
        }),
      ),
    )
  }

  /**
   * The rooms that get an autonomous round: not paused, not finished, active
   * within {@link ACTIVE_WINDOW_MS}, `world_id IS NULL` (a TRPG room's turns
   * come from the game tapes; scheduling one here would run chat-room agents
   * inside a world), newest first, capped, and holding {@link MIN_AGENTS} agents.
   * `= 0` is *false* for a NULL column in SQLite, so NULL flags mean skipped.
   *
   * The agent count is a second query rather than a join because the cap applies
   * to the *rooms* selected: a join would need a `GROUP BY` before the limit or
   * risk the limit counting membership rows.
   */
  activeRooms(): Room[] {
    const cutoff = new Date(Date.now() - ACTIVE_WINDOW_MS)

    const candidates = this.db
      .select()
      .from(rooms)
      .where(
        and(
          eq(rooms.isPaused, false),
          eq(rooms.isFinished, false),
          gte(rooms.lastActivityAt, cutoff),
          isNull(rooms.worldId),
        ),
      )
      .orderBy(desc(rooms.lastActivityAt))
      .limit(this.maxConcurrentRooms)
      .all()

    if (candidates.length === 0) return []

    const counts = new Map(
      this.db
        .select({ roomId: roomAgents.roomId, total: count() })
        .from(roomAgents)
        .where(
          inArray(
            roomAgents.roomId,
            candidates.map((room) => room.id),
          ),
        )
        .groupBy(roomAgents.roomId)
        .all()
        .map((row) => [row.roomId, row.total] as const),
    )

    return candidates.filter((room) => (counts.get(room.id) ?? 0) >= MIN_AGENTS)
  }

  /**
   * The five-minute sweep: expired cache entries, the stats line, and the
   * orchestrator's stale supersede stamps. No per-session lock sweep is needed —
   * `SessionPool` deletes its open-serializing promise when the open settles.
   */
  cleanup(): void {
    try {
      const cache = getCache()
      cache.cleanupExpired()
      cache.logStats()

      this.orchestrator.cleanupStaleEntries?.()
    } catch (error: unknown) {
      logger.exception('Error during cache cleanup', error)
    }
  }
}

// Cap on how many of the passed functions run at once. Per tick rather than
// shared across ticks, which is safe because a tick awaits everything it starts
// and an overrunning tick is skipped rather than overlapped.
function semaphore(limit: number): <T>(body: () => Promise<T>) => Promise<T> {
  if (limit <= 0) return (body) => body()

  let active = 0
  const waiting: Array<() => void> = []

  const release = (): void => {
    active--
    waiting.shift()?.()
  }

  return async <T>(body: () => Promise<T>): Promise<T> => {
    if (active >= limit) await new Promise<void>((resolve) => waiting.push(resolve))
    active++
    try {
      return await body()
    } finally {
      release()
    }
  }
}
